wicketd/src/update_planner.rs

```diff
@@ -0,0 +1,489 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use crate::artifacts::WicketdArtifactStore;
+use crate::mgs::make_mgs_client;
+use crate::update_events::UpdateEventFailureKind;
+use crate::update_events::UpdateEventKind;
+use crate::update_events::UpdateEventSuccessKind;
+use crate::update_events::UpdateStateKind;
+use anyhow::bail;
+use anyhow::ensure;
+use anyhow::Context;
+use buf_list::BufList;
+use debug_ignore::DebugIgnore;
+use gateway_client::types::SpIdentifier;
+use gateway_client::types::SpType;
+use gateway_client::types::SpUpdateStatus;
+use gateway_client::types::UpdateBody;
+use gateway_messages::SpComponent;
+use omicron_common::api::internal::nexus::KnownArtifactKind;
+use omicron_common::update::ArtifactId;
+use slog::error;
+use slog::info;
+use slog::o;
+use slog::Logger;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::net::SocketAddrV6;
+use std::sync::Arc;
+use std::sync::Mutex as StdMutex;
+use std::time::Duration;
+use std::time::Instant;
+use thiserror::Error;
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use uuid::Uuid;
+
+// These three types are mirrors of the HTTP
+// `UpdateState`/`UpdateEvent`/`UpdateLog`, but with the timestamps stored as
+// `Instant`s. This allows us to convert them to `Duration`s (i.e., ages) when
+// returning them to a caller.
+#[derive(Clone, Debug)]
+struct UpdateState {
+    timestamp: Instant,
+    kind: UpdateStateKind,
+}
+
+impl From<UpdateState> for crate::update_events::UpdateState {
+    fn from(state: UpdateState) -> Self {
+        Self { age: state.timestamp.elapsed(), kind: state.kind }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct UpdateEvent {
+    timestamp: Instant,
+    kind: UpdateEventKind,
+}
+
+impl From<UpdateEvent> for crate::update_events::UpdateEvent {
+    fn from(event: UpdateEvent) -> Self {
+        Self { age: event.timestamp.elapsed(), kind: event.kind }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct UpdateLog {
+    current: Option<UpdateState>,
+    events: Vec<UpdateEvent>,
+}
+
+impl From<UpdateLog> for crate::update_events::UpdateLog {
+    fn from(log: UpdateLog) -> Self {
+        Self {
+            current: log.current.map(Into::into),
+            events: log.events.into_iter().map(Into::into).collect(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct UpdatePlanner {
+    mgs_client: gateway_client::Client,
+    artifact_store: WicketdArtifactStore,
+    update_plan: StdMutex<Result<UpdatePlan, UpdatePlanError>>,
+    running_updates: Mutex<BTreeMap<SpIdentifier, JoinHandle<()>>>,
+    // Note: Our inner mutex here is a standard mutex, not a tokio mutex. We
+    // generally hold it only log enough to update its state or push a new
+    // update event into its running log; occasionally we hold it long enough to
+    // clone it.
+    update_logs: Mutex<BTreeMap<SpIdentifier, Arc<StdMutex<UpdateLog>>>>,
```

Comment:
So I'm trying to figure out if there's a reason `running_updates` and `update_logs` are stored separately. Could this just be something like:

```rust
struct UpdatePlanner {
    // ...
    sp_update_data: Mutex<BTreeMap<SpIdentifier, SpUpdateData>>,
}

struct SpUpdateData {
    running: Option<JoinHandle<()>>,
    logs: Arc<StdMutex<UpdateLog>>,
}
```

Reply:
Maybe! I landed where I did after a lot of churn, and your suggestion does look better. I'll give it a shot and see if I bump into anything unexpected.

Reply:
No problems making this change: [96135ee](https://github.com/oxidecomputer/omicron/pull/2375/commits/96135ee80d6f4d0ad3db830c600014da512b669a)